Add function for creating a CRS variable in files

mauzey1 opened the pull request:
Resolves #774 

This PR adds the function `cmor_set_crs` to CMOR.  This function is similar to `cmor_set_grid_mapping` as it takes a grid mapping name along with other attributes to create a dimensionless variable for that grid mapping.  The difference is that the variable will be named `crs` instead of the grid mapping name, and will have the option for a text attribute named `crs_wkt`.  The `crs_wkt` attribute is added by CMOR but is not validated as a Well-known Text (WKT) string.

durack1 reviewed (changes requested):
@mauzey1 this looks like a great start. Would it be possible to drop an ncdump output in the conversation for this PR so we can see what the output looks like - if that seems to match what we're expecting, let's merge this, get it into a nightly and get some feedback on the new feature

mauzey1 commented:
@durack1 Here's the header of the file made by the CRS test.

```
$ ncdump -h CMIP6/CMIP6/ISMIP6/PCMDI/PCMDI-test-1-0/piControl-withism/r3i1p1f1/Lmon/baresoilFrac/gn/v20250501/baresoilFrac_Lmon_PCMDI-test-1-0_piControl-withism_r3i1p1f1_gn_198001-198002.nc 
netcdf baresoilFrac_Lmon_PCMDI-test-1-0_piControl-withism_r3i1p1f1_gn_198001-198002 {
dimensions:
        time = UNLIMITED ; // (2 currently)
        y = 3 ;
        x = 4 ;
        bnds = 2 ;
        vertices = 4 ;
        strlen = 11 ;
variables:
        double time(time) ;
                time:bounds = "time_bnds" ;
                time:units = "days since 1980" ;
                time:calendar = "360_day" ;
                time:axis = "T" ;
                time:long_name = "time" ;
                time:standard_name = "time" ;
        double time_bnds(time, bnds) ;
        double y(y) ;
                y:units = "m" ;
                y:axis = "Y" ;
                y:long_name = "y coordinate of projection" ;
                y:standard_name = "projection_y_coordinate" ;
        double x(x) ;
                x:units = "m" ;
                x:axis = "X" ;
                x:long_name = "x coordinate of projection" ;
                x:standard_name = "projection_x_coordinate" ;
        int crs ;
                crs:grid_mapping_name = "lambert_conformal_conic" ;
                crs:standard_parallel = -20., 20. ;
                crs:longitude_of_central_meridian = 175. ;
                crs:latitude_of_projection_origin = 13. ;
                crs:false_easting = 8. ;
                crs:false_northing = 0. ;
                crs:crs_wkt = "Text for the CRS WKT" ;
        double latitude(y, x) ;
                latitude:standard_name = "latitude" ;
                latitude:long_name = "latitude" ;
                latitude:units = "degrees_north" ;
                latitude:missing_value = 1.e+20 ;
                latitude:_FillValue = 1.e+20 ;
                latitude:bounds = "vertices_latitude" ;
        double longitude(y, x) ;
                longitude:standard_name = "longitude" ;
                longitude:long_name = "longitude" ;
                longitude:units = "degrees_east" ;
                longitude:missing_value = 1.e+20 ;
                longitude:_FillValue = 1.e+20 ;
                longitude:bounds = "vertices_longitude" ;
        double vertices_latitude(y, x, vertices) ;
                vertices_latitude:units = "degrees_north" ;
                vertices_latitude:missing_value = 1.e+20 ;
                vertices_latitude:_FillValue = 1.e+20 ;
        double vertices_longitude(y, x, vertices) ;
                vertices_longitude:units = "degrees_east" ;
                vertices_longitude:missing_value = 1.e+20 ;
                vertices_longitude:_FillValue = 1.e+20 ;
        char type(strlen) ;
                type:long_name = "surface type" ;
                type:standard_name = "area_type" ;
        float baresoilFrac(time, y, x) ;
                baresoilFrac:standard_name = "area_fraction" ;
                baresoilFrac:long_name = "Bare Soil Percentage Area Coverage" ;
                baresoilFrac:comment = "no future, ISMIP6_table_comment: Percentage of entire grid cell  that is covered by bare soil." ;
                baresoilFrac:units = "%" ;
                baresoilFrac:history = "no history 2025-05-01T22:32:44Z altered by CMOR: Converted units from \'\' to \'%\'. 2025-05-01T22:32:44Z altered by CMOR: Treated scalar dimension: \'type\'. 2025-05-01T22:32:44Z altered by CMOR: Converted type from \'d\' to \'f\'." ;
                baresoilFrac:cell_methods = "area: mean where land over all_area_types time: mean" ;
                baresoilFrac:cell_measures = "area: areacella" ;
                baresoilFrac:coordinates = "type latitude longitude" ;
                baresoilFrac:missing_value = 1.e+20f ;
                baresoilFrac:_FillValue = 1.e+20f ;
                baresoilFrac:grid_mapping = "crs" ;

// global attributes:
                :Conventions = "CF-1.7 CMIP-6.2" ;
                :activity_id = "ISMIP6" ;
                :branch_method = "no parent" ;
                :branch_time_in_child = 59400. ;
                :branch_time_in_parent = 0. ;
                :contact = "Python Coder (coder@a.b.c.com)" ;
                :creation_date = "2025-05-01T22:32:44Z" ;
                :data_specs_version = "01.00.33" ;
                :experiment = "preindustrial control with interactive ice sheet" ;
                :experiment_id = "piControl-withism" ;
                :external_variables = "areacella" ;
                :forcing_index = 1 ;
                :frequency = "mon" ;
                :further_info_url = "https://furtherinfo.es-doc.org/CMIP6.PCMDI.PCMDI-test-1-0.piControl-withism.none.r3i1p1f1" ;
                :grid = "native atmosphere regular grid (3x4 latxlon)" ;
                :grid_label = "gn" ;
                :history = "2025-05-01T22:32:44Z ;rewrote data to be consistent with ISMIP6 for variable baresoilFrac found in table Lmon.;\n",
                        "Output from archivcl_A1.nce/giccm_03_std_2xCO2_2256." ;
                :initialization_index = 1 ;
                :institution = "Program for Climate Model Diagnosis and Intercomparison, Lawrence Livermore National Laboratory, Livermore, CA 94550, USA" ;
                :institution_id = "PCMDI" ;
                :mip_era = "CMIP6" ;
                :nominal_resolution = "10000 km" ;
                :parent_activity_id = "no parent" ;
                :parent_experiment_id = "no parent" ;
                :parent_mip_era = "no parent" ;
                :parent_source_id = "no parent" ;
                :parent_time_units = "no parent" ;
                :parent_variant_label = "no parent" ;
                :physics_index = 1 ;
                :product = "model-output" ;
                :realization_index = 3 ;
                :realm = "land" ;
                :references = "Model described by Koder and Tolkien (J. Geophys. Res., 2001, 576-591).  Also see http://www.GICC.su/giccm/doc/index.html.  The ssp245 simulation is described in Dorkey et al. \'(Clim. Dyn., 2003, 323-357.)\'" ;
                :run_variant = "3rd realization" ;
                :source = "PCMDI-test 1.0 (1989): \n",
                        "aerosol: none\n",
                        "atmos: Earth1.0-gettingHotter (360 x 180 longitude/latitude; 50 levels; top level 0.1 mb)\n",
                        "atmosChem: none\n",
                        "land: Earth1.0\n",
                        "landIce: none\n",
                        "ocean: BlueMarble1.0-warming (360 x 180 longitude/latitude; 50 levels; top grid cell 0-10 m)\n",
                        "ocnBgchem: none\n",
                        "seaIce: Declining1.0-warming (360 x 180 longitude/latitude)" ;
                :source_id = "PCMDI-test-1-0" ;
                :source_type = "AOGCM ISM AER" ;
                :sub_experiment = "none" ;
                :sub_experiment_id = "none" ;
                :table_id = "Lmon" ;
                :table_info = "Creation Date:(18 November 2020) MD5:2d2d31c26cc3dc3fb52409f4b0d9ae56" ;
                :title = "PCMDI-test-1-0 output prepared for CMIP6" ;
                :variable_id = "baresoilFrac" ;
                :variant_label = "r3i1p1f1" ;
                :license = "CMIP6 model data produced by Lawrence Livermore PCMDI is licensed under a Creative Commons Attribution 4.0 International License (https://creativecommons.org/licenses/by/4.0/). Consult https://pcmdi.llnl.gov/CMIP6/TermsOfUse for terms of use governing CMIP6 output, including citation requirements and proper acknowledgment. Further information about this data, including some limitations, can be found via the further_info_url (recorded as a global attribute in this file) and at https:///pcmdi.llnl.gov/. The data producers and data providers make no warranty, either express or implied, including, but not limited to, warranties of merchantability and fitness for a particular purpose. All liabilities arising from the supply of the information (including any liability arising in negligence) are excluded to the fullest extent permitted by law." ;
                :cmor_version = "3.10.0" ;
                :tracking_id = "hdl:21.14100/80b3b624-47b6-4c5a-ad4c-dc79681d7674" ;
}
```

durack1 commented:
Great, this looks like an excellent start.

```python
int crs ;
        crs:grid_mapping_name = "lambert_conformal_conic" ;
        crs:standard_parallel = -20., 20. ;
        crs:longitude_of_central_meridian = 175. ;
        crs:latitude_of_projection_origin = 13. ;
        crs:false_easting = 8. ;
        crs:false_northing = 0. ;
        crs:crs_wkt = "Text for the CRS WKT" ;
```

Following the defaults defined in the [CF: Appendix F: Grid Mappings](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#appendix-grid-mappings) to conform, did you want to also add in
```python
crs:earth_radius = 6378137.0 ;
crs:semi_major_axis = 6378137.0 ;
crs:longitude_of_prime_meridian = 0.0 ;
crs:prime_meridian_name = "Greenwich, London, UK"
crs:semi_major_axis = 6378137.0 ;
crs:inverse_flattening = 298.257223563 ;
crs:reference_ellipsoid_name = "WGS-84" ;
```
These appear to align well with the Reference entries at least for WGS-84, which appears to be a common latitude_longitude reference, It is consistent to the integer with the subsequent and latest IERS (2003) (see the historical table [here](https://en.wikipedia.org/wiki/Earth_ellipsoid#History)

@taylor13 @TomLav @DavidHassell to you folks have anything else to add?

TomLav commented:
Hei all,

This is a great start indeed ! I agree with @durack1 that adding the datum information is of interest. You'd probably have to check that the lat/lon actually correspond to the x/y/crs (projection + datum). e.g. what if the lat/lons are with a spherical Earth instead? It will depend on the dataset.

I maybe miss the x_bounds and y_bounds variables:
```
double y(y) ;
                y:units = "m" ;
                y:axis = "Y" ;
                y:long_name = "y coordinate of projection" ;
                y:standard_name = "projection_y_coordinate" ;
                y:bounds = "y_bnds";
double x(x) ;
                x:units = "m" ;
                x:axis = "X" ;
                x:long_name = "x coordinate of projection" ;
                x:standard_name = "projection_x_coordinate" ;
                x:bounds = "x_bnds";
double x_bnds(x, bnds);
double y_bnds(y, bnds);
```

In fact, we have clarified in the latest version of the CF convention that an application can be expected to compute the lat and lon vertices of each grid cell from the knowledge of x, y, x_bnds, and y_bnds (+ crs). I do not say that you should drop the lat, lon, vertices_longitude, and vertices_latitude in this file, but an application could bypass these and directly apply the x/y information combined with the crs instead.

Editorial: maybe have `latitude_vertices` instead of `vertices_latitude` by analogy to `time_bnds` ?

mauzey1 commented:
@durack1 For `polar_stereographic`, do we remove the deprecated `straight_vertical_longitude_from_pole` attribute and replace its use in tests with `longitude_of_projection_origin`?

taylor13 commented:
I support replacing the deprecated attributed as suggested.

durack1 commented:
> @durack1 For `polar_stereographic`, do we remove the deprecated `straight_vertical_longitude_from_pole` attribute and replace its use in tests with `longitude_of_projection_origin`?

@mauzey1 I also agree. If there are any other deprecations that the latest CF 1.12 captures, then I think we should catch those too - I didn't spot any others

durack1 commented:
If it wasn't obvious, the entries contained in the [#786 TestTables/CFgridMapping.json](https://github.com/PCMDI/cmor/pull/786/files#diff-d0142b6df01fc8e3c40806aa07f2578de43b78a44581147e654523b468c1ab03) file were an attempt of capturing the latest information from the http://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#appendix-grid-mappings details

mauzey1 commented:
I'll need to add the ability to use string attributes since the `cmor_set_grid_mapping` function only accepted numeric attributes.  It can be arrays of names and values like the numeric attributes.  It can also be used for the `crs_wkt` attribute instead of having its own parameter.

There's also the matter of the `towgs84` attribute, which is an array of up to 7 numbers similar to `standard_parallel`.  The way `standard_parallel` is handled by CMOR is by splitting it into `standard_parallel1` and `standard_parallel2` attributes passed to `cmor_set_grid_mapping`, and them combining them into `standard_parallel` when writing the file.  Should we do something similar with `towgs84`?

taylor13 commented:
> Should we do something similar with towgs84?

perhaps that's not a particularly elegant approach, but I would go with what is easiest if it won't be too confusing to the data writers.

durack1 commented:
> Should we do something similar with `towgs84`?

Thanks for pushing on this @mauzey1. I just tried to update myself as to what this is, following from Table F.1. in the CF 1.12 ([here](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#vertical-perspective)), we get:

This indicates a list of up to 7 Bursa Wolf transformation parameters, which can be used to approximate a transformation from the horizontal datum to the WGS84 datum. More precise datum transformations can be done with datum shift grids. Represented as a double-precision array, with 3, 6 or 7 values (if there are less than 7 values the remaining are considered to be zero). Corresponds to a OGC WKT TOWGS84 node.

Maybe we just leave it up to a user to correctly define this, it's an optional attribute anyway, right?

durack1 commented:
@davidhassell thanks for engaging, having another set of eyes on this is great!

The `crs_wkt` has been a hard one to figure out. Long story short, we were struggling to find ways to validate this, with numerous examples appearing to be completely counter-intuitive, but it's what other apparently valid data have included (example below). So for this reason, we thought we'd leave this string definition up to a user to get it right, with zero validation... If you have better ideas, we are all ears!

The example is this one below, which looks like it includes conflicting info to me!

```python
$ ncdump -h macav3_ACCESS-CM2_historical_2009.nc 
netcdf macav3_ACCESS-CM2_historical_2009 {
dimensions:
	crs = 1 ;
	time = UNLIMITED ; // (365 currently)
	lat = 671 ;
	lon = 1271 ;
variables:
	short crs(crs) ;
		crs:grid_mapping_name = "latitude_longitude" ;
		crs:longitude_of_prime_meridian = 0. ;
		crs:semi_major_axis = 6378137. ;
		crs:long_name = "WGS 84" ;
		crs:inverse_flattening = 298.257223563 ;
		crs:GeoTransform = "-179.5 0.1 0 74.5 0.1" ;
		crs:spatial_ref = "GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]]" ;
...
```

mauzey1 commented:
I've modified the `cmor_set_crs` function to now accept any kind of text attribute including `crs_wkt`.

CMOR with this feature will still not validate the value in `crs_wkt` as a valid WKT string.

durack1 commented:
> Editorial: maybe have `latitude_vertices` instead of `vertices_latitude` by analogy to `time_bnds` ?

@TomLav this is a good catch, I agree with you. I wonder if anything else could be caught and cleaned up. @taylor13 do you have a use case where `vertices_latitude` is the preferred over `latitude_vertices`

taylor13 commented:
regarding the naming of the bounds variables, this can be freely chosen; the CF conventions have not rules or guidance on this.   I think these variables were already in versions of CMOR used to write CMIP6 data on grids (other than longitudexlatitude grids), and they were given the names `vertices_longitude` and `vertices_latitude`.  I would stay consistent with how it was done in CMIP6.  [If it wasn't already in CMOR, please let me know.]

durack1 reviewed (approved):
@mauzey1 looks good to me, I think we're ready to move this across to a nightly build to get some feedback before a CMOR 3.11 final release is completed

@taylor13 @TomLav @davidhassell if there are any other loose ends I have not caught, we can capture additional tweaks in #774
